### PR TITLE
ANSI 301-2022: Shared batteries

### PR DESCRIPTION
**Pull request by shorowit**
## Pull Request Description

Allows shared batteries (batteries serving multiple dwelling units). Battery charging/discharging will be apportioned to the dwelling unit using its number of bedrooms divided by the total number of bedrooms served by the battery.

## Checklist

PR Author: Check these when they're done. Not all may apply. ~~strikethrough~~ and check any that do not apply. 

PR Reviewer: Verify each has been completed.

- [x] Schematron validator (`EPvalidator.xml`) has been updated
- [x] Sample files have been added/updated (`openstudio tasks.rb update_hpxmls`)
- [x] Tests have been added/updated (e.g., `HPXMLtoOpenStudio/tests/test*.rb` and/or `workflow/tests/test*.rb`)
- [x] Documentation has been updated
- [x] Changelog has been updated
- [x] `openstudio tasks.rb update_measures` has been run
- [x] No unexpected changes to simulation results of sample files

### CODE REVIEW

**Comment by afontani** [comment]
Is there any reason there couldn't be multiple shared batteries? I am thinking more about EVs than a stationary battery.

**Comment by shorowit** [comment]
> Is there any reason there couldn't be multiple shared batteries? I am thinking more about EVs than a stationary battery.

See [here](https://github.com/NREL/OpenStudio-HPXML/issues/1428). Currently the user has to combine multiple batteries into a single HPXML `Battery` element.

File: HPXMLtoOpenStudio/resources/battery.rb [hunk 1]
```diff
@@ -37,6 +37,15 @@ def self.apply(runner, model, pv_systems, battery, schedules_file, unit_multipli
...
     return if rated_power_output <= 0 || nominal_capacity_kwh <= 0 || battery.nominal_voltage <= 0
 
+    if battery.is_shared_system
+      # Apportion to single dwelling unit by # bedrooms
+      fail if battery.number_of_bedrooms_served.to_f <= nbeds.to_f # EPvalidator.xml should prevent this
+
+      nominal_capacity_kwh = nominal_capacity_kwh * nbeds.to_f / battery.number_of_bedrooms_served.to_f
+      usable_capacity_kwh = usable_capacity_kwh * nbeds.to_f / battery.number_of_bedrooms_served.to_f
+      rated_power_output = rated_power_output * nbeds.to_f / battery.number_of_bedrooms_served.to_f
+    end
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/battery.rb:47]: I took the new base-bldgtype-mf-unit-shared-pv-battery.xml sample file, temporarily changed the shared battery so that all inputs (NominalCapacity, UsableCapacity, RatedPowerOutput, and NumberofBedroomsServed) were multiplied by 10, and ran it. Simulation results were identical to the original file.

File: HPXMLtoOpenStudio/resources/battery.rb [hunk 2]
```diff
@@ -79,6 +88,7 @@ def self.apply(runner, model, pv_systems, battery, schedules_file, unit_multipli
     elcs.setNumberofStringsinParallel(number_of_strings_in_parallel)
     elcs.setInitialFractionalStateofCharge(0.0)
     elcs.setBatteryMass(battery_mass)
+    elcs.setDCtoDCChargingEfficiency(battery.round_trip_efficiency) # Note: This is currently unused in E+, so we use an EMS program below instead
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/battery.rb:91]: It was confusing to have the IDF show a round trip efficiency that doesn't match the HPXML inputs, so I added this in even though it doesn't affect the E+ results.

File: HPXMLtoOpenStudio/resources/hpxml_schema/HPXML.xsd [hunk 3]
```diff
@@ -2155,6 +2155,12 @@
 									<xs:element minOccurs="0" name="ModelNumber" type="Model"/>
 									<xs:element minOccurs="0" name="SerialNumber" type="HPXMLString"/>
 									<xs:element minOccurs="0" ref="ConnectedDevice"/>
+									<xs:element minOccurs="0" name="IsSharedSystem" type="HPXMLBoolean">
+										<xs:annotation>
+											<xs:documentation>Does the system serve multiple building/dwelling units?</xs:documentation>
+										</xs:annotation>
+									</xs:element>
+									<xs:element minOccurs="0" name="NumberofUnitsServed" type="HPXMLInteger"/>
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/hpxml_schema/HPXML.xsd:2163]: New elements from https://github.com/hpxmlwg/hpxml/pull/397.

File: HPXMLtoOpenStudio/resources/hpxml_schematron/EPvalidator.xml [hunk 4]
```diff
@@ -2009,6 +2010,15 @@
     </sch:rule>
   </sch:pattern>
 
+  <sch:pattern>
+    <sch:title>[BatteryType=Shared]</sch:title>
+    <sch:rule context='/h:HPXML/h:Building/h:BuildingDetails/h:Systems/h:Batteries/h:Battery[h:IsSharedSystem="true"]'>
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/hpxml_schematron/EPvalidator.xml:2015]: Copied the pattern from shared PV systems.

File: HPXMLtoOpenStudio/tests/test_battery.rb [hunk 5]
```diff
@@ -125,7 +125,43 @@ def test_pv_battery
...
+      assert_equal(6000.0, elcd.designStorageControlDischargePower.get)
+      assert(!elcd.demandLimitSchemePurchasedElectricDemandLimit.is_initialized)
+      assert_equal('TrackFacilityElectricDemandStoreExcessOnSite', elcd.storageOperationScheme)
+      assert(!elcd.storageChargePowerFractionSchedule.is_initialized)
+      assert(!elcd.storageDischargePowerFractionSchedule.is_initialized)
+      assert(!elcd.storageConverter.is_initialized)
+    end
+  end
+
+  def test_pv_battery_shared
```
Comment (shorowit) [inline HPXMLtoOpenStudio/tests/test_battery.rb:153]: New test.

File: ReportSimulationOutput/measure.rb [hunk 6]
```diff
@@ -1035,9 +1037,9 @@ def get_outputs(runner, args, building_id)
       batt_loss = get_report_variable_data_timeseries(keys, vars, UnitConversions.convert(1.0, 'J', 'kWh'), 0, resilience_frequency)
 
       min_soc = elcd.minimumStorageStateofChargeFraction
-      batt_kw = battery.rated_power_output / 1000.0
-      batt_roundtrip_eff = battery.round_trip_efficiency
-      batt_kwh = Battery.get_usable_capacity_kWh(battery)
+      batt_kw = elcd.designStorageControlDischargePower.get / 1000.0
+      batt_roundtrip_eff = elcs.dctoDCChargingEfficiency
+      batt_kwh = elcs.additionalProperties.getFeatureAsDouble('UsableCapacity_kWh').get
```
Comment (shorowit) [inline ReportSimulationOutput/measure.rb:1042]: For the reslience output metric, now getting values from the OpenStudio objects rather than the HPXML Battery object so I don't have to duplicate the prorating equations from battery.rb here.

File: HPXMLtoOpenStudio/tests/test_battery.rb (hunk 5 above)
Comment (jmaguire1) [reply HPXMLtoOpenStudio/tests/test_battery.rb:153]: Also good to change the default efficiency to match our OS-HPXML, good catch that it's not actually 0.95.

**Review by jmaguire1 — COMMENTED** [review]
Looks pretty good Scott, nice job! One suggestion: I didn't see a check anywhere that number of bedrooms served has to be <= total number of bedrooms. Did I miss it, or if I didn't should we add one? It's an edge case where a user might be providing bad inputs.

**Comment by shorowit** [comment]
> Looks pretty good Scott, nice job! One suggestion: I didn't see a check anywhere that number of bedrooms served has to be <= total number of bedrooms. Did I miss it, or if I didn't should we add one? It's an edge case where a user might be providing bad inputs.

Thanks, Jeff. It is indeed checked in the Schematron validation (EPvalidator.rb) and stated in the documentation, though I just simplified the documentation to make it more obvious.

**Comment by jmaguire1** [comment]
Ah I figured I must have just missed it, but now I see it. Looks good then!
